### PR TITLE
[WIP] fixes #26886 - graphql: discovery rules

### PR DESCRIPTION
**Pull request by timogoebel**

### CODE REVIEW

**Review by ares — COMMENTED** [review]
Before we go ahead with plugins, is there some plugin interface / dsl that could abstract this? It can also be added later, but before we copy the pattern everywhere...

**Comment by timogoebel** [comment]
> is there some plugin interface / dsl that could abstract this?

I created https://projects.theforeman.org/issues/26541 a while ago to implement this.
I was thinking of a DSL like this:

```ruby
extend_graphql_type type: ::Types::Query do
  record_field :discovery_rule, Types::DiscoveryRule
  collection_field :discovery_rules, Types::DiscoveryRule
end
```

and

```ruby
extend_graphql_type type: Types::Query,
  with_module: ::Types::Extensions::ForemanDiscovery::QueryExtensions
```

**Comment by timogoebel** [comment]
WIP-PR for ^^: https://github.com/theforeman/foreman/pull/6809

**Comment by xprazak2** [comment]
https://github.com/theforeman/foreman/pull/6866 should fix the test failures

**Comment by xprazak2** [comment]
[test foreman_discovery]

**Comment by lzap** [comment]
Can you try again? I fixed tests.

**Comment by timogoebel** [comment]
@kamils-iRonin: Do you mind reviewing this? Am I using the latest APIs here?

File: lib/foreman_discovery/engine.rb [hunk 1]
```diff
@@ -215,6 +216,9 @@ class Engine < ::Rails::Engine
       # Controller extensions
       ::HostsController.send :include, ForemanDiscovery::Concerns::HostsControllerExtensions
       ::Api::V2::FactValuesController.send :include, Api::V2::FactValuesControllerExtensions
+
+      # Graphql extensions
+      Types::Query.send(:include, Types::Extensions::ForemanDiscovery::QueryExtensions)
```
Comment (kamils-iRonin) [inline lib/foreman_discovery/engine.rb:221]: I think you can use [register_graphql_query_field](https://github.com/theforeman/foreman/blob/ed3abb68ad71c365ab5f0992ec543234e1376a6a/app/registries/foreman/plugin.rb#L561-L563) 

```ruby
register_graphql_query_field : discovery_rule, 'Types::DiscoveryRule', :record_field
register_graphql_query_field : discovery_rules, 'Types::DiscoveryRule', :collection_field
```

File: app/graphql/types/discovery_rule.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+module Types
```
Comment (kamils-iRonin) [inline app/graphql/types/discovery_rule.rb:1]: I would move it into `ForemanDiscovery` namespace, so `ForemanDiscovery::Types::DiscoveryRule`

**Comment by stejskalleos** [comment]
I'm closing the PR since the last update was on Oct 12, 2020.

**Pull request closed without merging**